torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -0,0 +1,295 @@
+#include "ProcessGroupMPI.hpp"
+
+#include <map>
+#include <iostream>
+
+namespace c10d {
+
+#define MPI_CHECK(cmd) do {                                \
+  int mpiStatus = cmd;                                     \
+  if (mpiStatus != MPI_SUCCESS) {                          \
+    std::string err = "MPI error in: " +                   \
+                      std::string(__FILE__) + ":" +        \
+                      std::to_string(__LINE__) +           \
+                      ", with error code: " +              \
+                      std::to_string(mpiStatus);           \
+    std::cerr << err << std::endl;                         \
+    throw std::runtime_error(err);                         \
+  }                                                        \
+} while (0)
+
+namespace {
+
+// Op mapping
+std::map<ReduceOp, MPI_Op> mpiOp = {
+  {ReduceOp::MIN, MPI_MIN},
+  {ReduceOp::MAX, MPI_MAX},
+  {ReduceOp::SUM, MPI_SUM},
+  {ReduceOp::PRODUCT, MPI_PROD},
+};
+// Type mapping
+std::map<at::ScalarType, MPI_Datatype> mpiDatatype = {
+  {at::kByte, MPI_UNSIGNED_CHAR},
+  {at::kChar, MPI_CHAR},
+  {at::kDouble, MPI_DOUBLE},
+  {at::kFloat, MPI_FLOAT},
+  {at::kInt, MPI_INT},
+  {at::kLong, MPI_LONG},
+  {at::kShort, MPI_SHORT},
+};
+
+// Checking CUDA-aware MPI support
+bool cudaAwareMpiCheck() {
+#if defined(MPIX_CUDA_AWARE_SUPPORT) && MPIX_CUDA_AWARE_SUPPORT
+// Compile time check
+#elif defined(MPIX_CUDA_AWARE_SUPPORT) && !MPIX_CUDA_AWARE_SUPPORT
+  return false;
+#else
+  return false;
+#endif /* MPIX_CUDA_AWARE_SUPPORT */
+
+// Run time check
+#if defined(MPIX_CUDA_AWARE_SUPPORT)
+  if (MPIX_Query_cuda_support() == 1) {
+    return true;
+  } else {
+    return false;
+  }
+#else /* !defined(MPIX_CUDA_AWARE_SUPPORT) */
+  return false;
+#endif /* MPIX_CUDA_AWARE_SUPPORT */
+}
+
+} // namespace
+
+
+// ProcessGroupMPI::WorkMPI
+ProcessGroupMPI::WorkMPI::WorkMPI() : completed_(false) {}
+
+ProcessGroupMPI::WorkMPI::~WorkMPI() {}
+
+bool ProcessGroupMPI::WorkMPI::isCompleted() const {
+  return completed_;
+}
+
+bool ProcessGroupMPI::WorkMPI::isSuccess() const {
+  return !ex_;
+}
+
+bool ProcessGroupMPI::WorkMPI::wait() {
+  std::unique_lock<std::mutex> lock(workMutex_);
+  while (!completed_) {
+    workCV_.wait(lock);
+  }
+  return isSuccess();
+}
+
+void ProcessGroupMPI::WorkMPI::finish() {
+  {
+    std::unique_lock<std::mutex> lock(workMutex_);
+    completed_ = true;
+  }
+  workCV_.notify_all();
+}
+
+void ProcessGroupMPI::WorkMPI::finishWithException(
+    const std::exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(workMutex_);
+    completed_ = true;
+    ex_ = std::unique_ptr<std::exception>(new std::exception(ex));
```

Comment:
You should use `exception_ptr` here. Doing `new std::exception(ex)` will do a copy-construction of `std::exception` and might cause struct slicing

Reply:
changed